### PR TITLE
Added join command.

### PR DESCRIPTION
**Pull request by cseed**

### CODE REVIEW

File: src/main/scala/org/broadinstitute/hail/driver/Join.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      fatal(
+        s"""cannot join datasets with different sample schemata
+            |  left sample schema: @1
+            |  right sample schema: @2""".
+          stripMargin,
+        left.saSignature.toPrettyString(compact = true, printAttrs = true),
+        right.saSignature.toPrettyString(compact = true, printAttrs = true))
+    }
+
+    val joined = left.rdd.orderedLeftJoinDistinct(right.rdd)
```
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/driver/Join.scala]: Prefer to put this on OrderedRDD:
```
  def orderedInnerJoinDistinct[V2](other: OrderedRDD[PK, K, V2]): RDD[(K, (V, V2))] =
    orderedLeftJoinDistinct(other).flatMapValues { case (v1, v2) => v2.map((v1, _)) }
```

File: src/main/scala/org/broadinstitute/hail/driver/Join.scala [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      .flatMapValues { case ((lva, lgs), None) =>
+        None
+      case ((lva, lgs), Some((rva, rgs))) =>
+        Some((lva, lgs ++ rgs))
+      }
+      .toOrderedRDD
```
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/driver/Join.scala]: Should be `asOrderedRDD`, not `toOrderedRDD`

File: src/main/scala/org/broadinstitute/hail/driver/Join.scala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  def supportsMultiallelic = true
+
+  def requiresVDS = true
+
+  def run(state: State, options: Options): State = {
+    val left = state.vds
+    val right = VariantSampleMatrix.read(state.sqlContext, options.right)
+
+    if (left.wasSplit != right.wasSplit) {
+      fatal(
```
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/driver/Join.scala]: Unsure if this needs to be a fatal -- inner join means that it's safe, as long as we document it.  Perhaps a warning?

File: src/main/scala/org/broadinstitute/hail/driver/Join.scala [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    val newRightSampleIds = right.sampleIds.map { s =>
+      var s2 = s
+      var changed = false
+      while (sampleSet.contains(s2)) {
+        s2 = s2 + "d"
+        changed = true
+      }
+      if (changed)
+        info(s"renamed right sampleId $s to $s2")
```
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/driver/Join.scala]: Could get pretty verbose.  Maybe we can use a truncatable and write the full thing to the log?

<  build up `mapping: Array[(original, changedName)]`  >
```
if (mapping.length > 0)
  info(s""" Renamed ${mapping.length} ids:  
    |  @1""".stripMargin, mapping
    .map { case (pre,post) => s"$pre => $post" }.truncatable("\n  "))

File: src/main/scala/org/broadinstitute/hail/driver/Join.scala [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    val right = VariantSampleMatrix.read(state.sqlContext, options.right)
+
+    if (left.wasSplit != right.wasSplit) {
+      fatal(
+        s"""cannot join split and unsplit datasets
+            |  left was split: ${ left.wasSplit }
+            |  light was split: ${ right.wasSplit }""".stripMargin)
+    } else if (left.saSignature != right.saSignature) {
+      fatal(
+        s"""cannot join datasets with different sample schemata
```
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/driver/Join.scala]: might want option to use left sample schema, and set all right sample annotations to missing

File: src/main/scala/org/broadinstitute/hail/driver/Join.scala [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            |  left sample schema: @1
+            |  right sample schema: @2""".
+          stripMargin,
```
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/driver/Join.scala]: weird formatting

File: src/main/scala/org/broadinstitute/hail/driver/Join.scala (hunk 5 above)
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/driver/Join.scala]: I'm inclined to make the join as simple as possible.  You can do that with an `annotatesamples` easily enough.

**Pull request merged**